Add a switch for fixed strings on input

forejtv opened the pull request:
This adds a `--string-input-value` switch for jbmc that allows to specify one or more fixed values that are considered for input strings, instead of an unbounded range of (possibly non-sensical) strings whose choice is at a mercy of the SAT solver.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 67e1a6b29274828e7ea818d094d318f113edc309).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93866539

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 6f399efd16437c9b8c6289c91bd5adea599dd331).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93876986
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1f16d74a7c9236f336ee41f0191b05c15d67ee45).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93899261

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1009,29 +1005,40 @@ void java_object_factoryt::gen_nondet_struct_init(
 
-    // If the initialised type is a special-cased String type (one with length
-    // and data fields introduced by string-library preprocessing), initialise
-    // those fields with nondet values:
```
I think it's better to keep this comment

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1009,29 +1005,40 @@ void java_object_factoryt::gen_nondet_struct_init(
+    }
+    // in the else case we have
+    // '!is_char_sequence || !has_data_length'
+    // and so we are not dealing with a string
```
this comment is not useful

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1009,29 +1005,40 @@ void java_object_factoryt::gen_nondet_struct_init(
     // This code mirrors the `remove_java_new` pass:
-    auto initial_object = zero_initializer(struct_type, source_locationt(), ns);
+    auto initial_object =
+      zero_initializer(struct_type, source_locationt(), ns);
```
:pick: don't introduce reformatting if you don't touch these lines

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1009,29 +1005,40 @@ void java_object_factoryt::gen_nondet_struct_init(
   {
     class_identifier = struct_tag;
 
+    bool is_char_sequence =
```
const

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1009,29 +1005,40 @@ void java_object_factoryt::gen_nondet_struct_init(
+    bool is_char_sequence =
+      java_string_library_preprocesst
+        ::implements_java_char_sequence(struct_type);
+    bool has_data_length =
```
const and :pick: rename `has_length_and_data`

romainbrenguier commented on jbmc/regression/jbmc-strings/string-input-value/Test.java:
```diff
@@ -0,0 +1,12 @@
+
+    static void checkStringInputValue(String s) {
+      if (s.length() == 2)
+        assert(false);
```
:pick: use `{` `}` on all `if` blocks

romainbrenguier commented on jbmc/regression/jbmc-strings/string-input-value/Test.java:
```diff
@@ -0,0 +1,12 @@
+        assert(false);
+      else if (s.length() == 4)
+        assert(false);
+      assert(false);
```
Also add a test where you check for the data in the string

romainbrenguier commented on jbmc/regression/jbmc-strings/string-input-value/Test.java:
```diff
@@ -0,0 +1,12 @@
+public class Test {
```
The test could be added in a separate commit

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
       struct_type.has_component("length") && struct_type.has_component("data");
+    // following is true if we were not given fixed string input values
+    // using --string-input-value
+    bool no_input_values =
```
const

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
     bool has_data_length =
       struct_type.has_component("length") && struct_type.has_component("data");
+    // following is true if we were not given fixed string input values
+    // using --string-input-value
```
:pick: this comment seems superfluous :tomato:

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
-      to_struct_expr(*initial_object), ns, symbol_typet(qualified_clsid));
+    if(!is_char_sequence || !has_data_length || no_input_values)
+    {
+      // Either not a string, or user did not specify --string-input-values
```
:tomato:

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
+      auto initial_object =
+        zero_initializer(struct_type, source_locationt(), ns);
+      CHECK_RETURN(initial_object.has_value());
+      irep_idt qualified_clsid = "java::" + id2string(class_identifier);
```
const

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
-    // in the else case we have
-    // '!is_char_sequence || !has_data_length'
-    // and so we are not dealing with a string
+    else
```
This function is too long, please don't make it longer

forejtv replied:
I have tried to put some lines to a separate function, but I don't see a simple way of not making the function longer. To be fair, quite a few of the lines are taken by comments :-)

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
+      // of one of the fixed input strings to the input variable in question
+
+      // the following invariant is due to no_input_values==false
+      auto size = object_factory_parameters.string_input_values.size();
```
const

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
+        exprt name_literal(ID_java_string_literal);
+        name_literal.set(ID_value, val);
+        symbol_exprt s =
+          get_or_create_string_literal_symbol(name_literal, symbol_table, true);
```
const

romainbrenguier commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -252,11 +252,22 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
+  {
+    warning() << "--string-input-value ignored due to --no-refine-strings"
+              << eom;
+  }
```
:pick: It would make more sense to have this check just before we set the option

romainbrenguier commented on src/solvers/refinement/string_refinement.h:
```diff
@@ -33,13 +33,15 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
   " --no-refine-strings          turn off string refinement\n" \
   " --string-printable           restrict to printable strings (experimental)\n" /* NOLINT(*) */ \
   " --string-non-empty           restrict to non-empty strings (experimental)\n" /* NOLINT(*) */ \
+  " --string-printable st        restrict non-null strings to a fixed value(s)\n" /* NOLINT(*) */ \
```
the help message should mention `st`, and should be made clearer for the multiple values case

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
+      }
+      // in the else case we have
+      // '!is_char_sequence || !has_data_length'
+      // and so we are not dealing with a string
```
:tomato:

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1012,75 @@ void java_object_factoryt::gen_nondet_struct_init(
-    irep_idt qualified_clsid = "java::" + id2string(class_identifier);
-    set_class_identifier(
-      to_struct_expr(*initial_object), ns, symbol_typet(qualified_clsid));
+    if(!is_char_sequence || !has_data_length || no_input_values)
```
:pick: I think doing the special case first would be clearer

romainbrenguier reviewed (changes requested):
Needs more tests and some refactoring to not make gen_nondet_struct_init longer when it is already more than 100 lines. I suggest putting the code you want to add in a helper function.

Commit:
Add a switch for fixed strings on input

Commit:
Test for --string-input-value

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fdc941c1989cca7013b603542f7961df72ccbb2e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94099113

peterschrammel commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -250,11 +250,13 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
-  if(cmdline.isset("no-refine-strings") && cmdline.isset("string-printable"))
   {
-    warning() << "--string-printable ignored due to --no-refine-strings" << eom;
+    if(cmdline.isset("no-refine-strings"))
```
:warning: There are 3 independent refactorings in this commit. Split them.

forejtv replied:
Good point. I've split this into two commits, since I think the two refactorings related to initialize_nondet_string_fields are too closely related to be done separately: one changes the caller to handle the informal preconditions, and one changes the callee to turn informal preconditions to proper preconditions.

peterschrammel commented on jbmc/src/java_bytecode/java_entry_point.cpp:
```diff
@@ -131,6 +131,21 @@ static void java_static_lifetime_init(
 
+  // If there are strings given using --string-input-value, we need to add
+  // them to the symbol table now, so that they appear in __CPROVER_initialize
+  // and we can refer to them later when we initialise input values.
```
:pick: initialize

peterschrammel commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1039,62 @@ void java_object_factoryt::gen_nondet_struct_init(
+    {
+      // Add an initial all-zero write. Most of the fields of this will be
+      // overwritten, but it helps to have a whole-structure write that analysis
+      // passes can easily recognise leaves no uninitialised state behind.
```
I struggle to understand this sentence.

forejtv replied:
Me too :-) The sentence was there before, I've only put it in an else block and hence indentation changed. I don't know how to improve the sentence since I don't have good enough understanding of the code.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fe741373ae1ed1c9ee1f98d5e0b55b96c1e732b4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94119132

Commit:
Move string-printable and no-refine-strings compatibility check.

Just moving it by a few lines to make the code easier to follow.

Commit:
Check initialize_nondet_string_fields's applicability before it is called.

Before, initialize_nondet_string_fields would check itself if the argument is a string,
and would return true/false depending on whether any change was applied.

Now the function asserts in precondition that the arguments are applicable, and the
caller is responsible for checking this.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f42b870b1faf92bdb042ba8ad5d8dafb36317e09).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94121222

forejtv commented:
@romainbrenguier @peterschrammel This is now ready to re-review.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 82a09fd2ab90478bbb4db94b72a78fcc3f67d727).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94127897

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -564,7 +564,7 @@ static mp_integer max_value(const typet &type)
 /// Unit tests in `unit/java_bytecode/java_object_factory/` ensure
 /// it is the case.
-bool initialize_nondet_string_fields(
+void initialize_nondet_string_fields(
```
Update the documentation

romainbrenguier commented on jbmc/src/java_bytecode/java_entry_point.cpp:
```diff
@@ -131,6 +131,21 @@ static void java_static_lifetime_init(
+  for(const auto &val : object_factory_parameters.string_input_values)
+  {
+    exprt my_literal(ID_java_string_literal);
+    my_literal.set(ID_value, val);
```
:bulb: ideally there would be a `java_string_literal_exprt` type

romainbrenguier commented on jbmc/regression/jbmc-strings/string-input-value/test1.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+Test.class
+--verbosity 10 --string-input-value fo --string-input-value barr --function Test.checkStringInputValue --string-input-value "" --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
```
:pick: Could we also have a test.desc where "" (and maybe others) are not given as string-input-value to see the difference?

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1010,37 +1039,62 @@ void java_object_factoryt::gen_nondet_struct_init(
+      // We create a switch statement where each case is an assignment
+      // of one of the fixed input strings to the input variable in question
+
+      alternate_casest cases =
```
const

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0ad4ec168afdfae06c3e6e8e0532c7d83f0ae6d1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/94254247
